frontends/election-manager/src/components/save_file_to_usb.tsx

```diff
@@ -32,7 +32,7 @@ export enum FileType {
 
 export interface Props {
   onClose: () => void;
-  generateFileContent: () => Promise<Uint8Array | string>;
+  generateFileContent: () => Uint8Array | string | Promise<Uint8Array | string>;
```

Comment:
This confused me. How do you call a function that may or may not return a promise? Seems like it might be easier to just have it always return a promise

Reply:
It's basically saying "I take a function that can be either synchronous or asynchronous". Perhaps this would be clearer with a type like this?

```ts
type MaybePromise<T> = T | Promise<T>;

generateFileContent: () => MaybePromise<Uint8Array | string>;
```

Reply:
Yeah I like that idea.

Also found this in MDN docs which I didn't know before:
> If the value of the expression following the await operator is not a Promise, it's converted to a [resolved Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve).

So effectively the return value of this function will end up being wrapped in a Promise, since the caller will have to `await` it.

I don't know if it's simpler to do the `MaybePromise` approach or just have a `Promise` return type and then either disable the lint rule for non-Promise functions or explicitly return a Promise. What do you think would be the clearest to readers (since functionally they're all the same)?

Reply:
I'm not sure which is better. I'm inclined toward the `MaybePromise` approach for situations where you really don't know whether the result will be async or not. For example, it'd be annoying if the callback to Jest's `test` had to always return a promise just because it can sometimes return a promise. For the mocking/faking cases, I think it makes sense to make the interface use async but then disable the lint rule for those specific cases, since mocks are not a typical use of the API.

I'll try out the `MaybePromise` thing and see how it feels.

Reply:
Sounds good!

Reply:
I ended up calling it `PromiseOr<T>`. What do you think?

Reply:
I usually read `Container<T>` as "Container of T" in my mind, so it reads a bit weird to me: "Promise Or of T". I liked `MaybePromise` (or `OptionalPromise` or something in that vein)

Reply:
I'd support either of those alternatives. One advantage `PromiseOr` has is that it'll auto-complete when typing `Promise`, which can aid discovery. But that's not a huge deal. I'm not sure whether `OptionalPromise` would be confused with `Optional`, i.e. would you expect `type OptionalPromise<T> = Promise<T> | undefined`? `MaybePromise` has a similar problem. Or maybe I'm overthinking it 😜

Reply:
We're definitely deep in the weeds haha. I agree `OptionalPromise` is more likely to be confused with `Optional`